File: backend/services/agent_engine.py
"""
Server-side Agent Engine with SSE (Server-Sent Events) streaming.
Runs the 7-phase pipeline and streams log events to the frontend.
"""

import asyncio
import json
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from database.models import Chemical, Catalyst, MsdsData, Simulation, DEMO_DATASET_ID
from services.optimizer import optimize_formulation
from services.gemini_service import generate_sop, generate_report


def _ts():
    return datetime.now().strftime("%H:%M:%S.%f")[:-3]


def _event(event_type: str, data: dict) -> str:
    """Format an SSE event."""
    return f"event: {event_type}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"


async def run_agent_stream(params: dict, db: AsyncSession):
    """
    Async generator that yields SSE events as the agent pipeline runs.
    Each yield is a formatted SSE string.
    """
    hazard_rank = {"low": 1, "medium": 2, "high": 3}
    max_hazard_rank = hazard_rank.get(params.get("maxHazard", "medium"), 2)
    dataset_id = params.get("datasetId") or DEMO_DATASET_ID

    # ── Phase 1: Input Parsing ─────────────────────────────
    yield _event("phase", {"phase": "parsing"})
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Agent initialized. Starting autonomous design pipeline..."})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}] Parsing input requirements..."})
    await asyncio.sleep(0.2)
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Target Heat Resistance: {params['heatResistance']}°C"})
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Target Tensile Strength: {params['tensileStrength']} MPa"})
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Min Eco Score: {params['ecoScore']}"})
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Application Domain: {params['domain']}"})
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   └─ Max Hazard Level: {params['maxHazard']}"})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ Input validation passed."})

    # ── Phase 2: Database Scan ─────────────────────────────
    yield _event("phase", {"phase": "scanning"})
    await asyncio.sleep(0.4)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 2: Scanning chemical database..."})

    # Query chemicals filtered by dataset
    chem_result = await db.execute(
        select(Chemical).where(Chemical.dataset_id == dataset_id)
    )
    all_chemicals = chem_result.scalars().all()
    chemicals_data = [
        {
            "id": c.id, "name": c.name, "name_ko": c.name_ko,
            "category": c.category, "sub_category": c.sub_category,
            "molecular_weight": c.molecular_weight, "heat_resistance": c.heat_resistance,
            "tensile_strength": c.tensile_strength, "elongation": c.elongation,
            "eco_score": c.eco_score, "density": c.density, "hazard_level": c.hazard_level,
            "cost_per_kg": c.cost_per_kg, "compatible_with": c.compatible_with or [],
            "color": c.color,
        }
        for c in all_chemicals
    ]

    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] Loaded {len(chemicals_data)} materials from database."})
    await asyncio.sleep(0.3)

    # Filter candidates
    candidates = [
        c for c in chemicals_data
        if hazard_rank.get(c["hazard_level"], 1) <= max_hazard_rank and c["category"] != "Additive"
    ]

    yield _event("log", {"type": "THINK", "text": f"[{_ts()}] Filtering by hazard constraint (≤ {params['maxHazard']})..."})
    await asyncio.sleep(0.2)
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ {len(candidates)} candidate materials passed filter."})

    for i, c in enumerate(candidates[:5]):
        sep = "└─" if i == min(4, len(candidates) - 1) else "├─"
        yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   {sep} [{c['id']}] {c['name']} | Heat: {c['heat_resistance']}°C | Str: {c['tensile_strength']} MPa"})
        await asyncio.sleep(0.1)

    # ── Phase 3: Compatibility Analysis ────────────────────
    yield _event("phase", {"phase": "analyzing"})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 3: Catalyst-polymer compatibility analysis..."})

    cat_result = await db.execute(
        select(Catalyst).where(Catalyst.dataset_id == dataset_id)
    )
    all_catalysts = cat_result.scalars().all()
    catalysts_data = [
        {
            "id": ct.id, "name": ct.name, "type": ct.type,
            "activation_temp": ct.activation_temp, "selectivity": ct.selectivity,
            "toxicity": ct.toxicity, "lifetime": ct.lifetime, "cost_per_kg": ct.cost_per_kg,
            "effect_on_strength": ct.effect_on_strength, "effect_on_heat_res": ct.effect_on_heat_res,
            "effect_on_eco": ct.effect_on_eco, "description": ct.description,
        }
        for ct in all_catalysts
    ]

    compat_catalysts = [
        ct for ct in catalysts_data
        if any(ct["id"] in (c.get("compatible_with") or []) for c in candidates)
    ]

    for ct in compat_catalysts[:4]:
        cc = sum(1 for c in candidates if ct["id"] in (c.get("compatible_with") or []))
        yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ {ct['name']}: compatible with {cc}/{len(candidates)} | Selectivity: {ct['selectivity']}%"})
        await asyncio.sleep(0.15)

    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ {len(compat_catalysts)} compatible catalysts identified."})

    # ── Phase 4: MSDS Safety Review ────────────────────────
    yield _event("phase", {"phase": "safety"})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 4: MSDS safety review..."})

    msds_result = await db.execute(select(MsdsData))
    all_msds = msds_result.scalars().all()
    msds_map = {m.chemical_id: m for m in all_msds}

    warnings = []
    for c in candidates:
        msds = msds_map.get(c["id"])
        if msds:
            if msds.signal_word == "Danger":
                warnings.append({
                    "materialId": c["id"],
                    "materialName": c["name"],
                    "signalWord": msds.signal_word,
                    "hazards": (msds.hazard_statements or [])[:2],
                    "protection": msds.personal_protection,
                })
                yield _event("log", {"type": "WARN", "text": f"[{_ts()}] ⚠ MSDS ALERT: {c['name']} — Signal Word: DANGER"})
                yield _event("log", {"type": "WARN", "text": f"[{_ts()}]   └─ Required PPE: {msds.personal_protection}"})
                await asyncio.sleep(0.15)
            else:
                yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ {c['name']}: MSDS reviewed — {msds.hazard_classification}"})
                await asyncio.sleep(0.08)

    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ Safety review complete. {len(warnings)} alert(s) flagged."})

    # ── Phase 5: Optimization ──────────────────────────────
    yield _event("phase", {"phase": "optimizing"})
    await asyncio.sleep(0.4)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 5: Running multi-objective optimization..."})
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] > python3 sim_optimize.py --mode=pareto --generations=500"})
    await asyncio.sleep(0.3)

    optimized = optimize_formulation(params, candidates, compat_catalysts)

    yield _event("log", {"type": "THINK", "text": f"[{_ts()}] Weighted Euclidean distance minimization..."})
    await asyncio.sleep(0.2)
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Gen 100/500: best_fitness = {optimized['score']*0.7:.4f}"})
    await asyncio.sleep(0.15)
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   ├─ Gen 300/500: best_fitness = {optimized['score']*0.9:.4f}"})
    await asyncio.sleep(0.15)
    yield _event("log", {"type": "THINK", "text": f"[{_ts()}]   └─ Gen 500/500: best_fitness = {optimized['score']:.4f} ★ converged"})
    await asyncio.sleep(0.2)

    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}] ═══════════════════════════════════════"})
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}] ★ Optimal formulation found (score: {optimized['score']:.1f}/100)"})
    for f in optimized["formulation"]:
        yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   ├─ {f['name']}: {f['ratio']:.1f}%"})
        await asyncio.sleep(0.08)
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   └─ Catalyst: {optimized['selectedCatalyst']['name']}"})

    # ── Phase 6: Property Prediction ───────────────────────
    yield _event("phase", {"phase": "simulating"})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 6: Predicting composite properties..."})
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] > python3 property_predictor.py --model=neural_net_v3"})
    await asyncio.sleep(0.4)

    pred = optimized["predictedProperties"]
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}] ── Predicted Composite Properties ──"})
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   ├─ Heat Resistance: {pred['heatResistance']}°C (target: {params['heatResistance']}°C)"})
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   ├─ Tensile Strength: {pred['tensileStrength']} MPa (target: {params['tensileStrength']} MPa)"})
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   ├─ Eco Score: {pred['ecoScore']}/100"})
    yield _event("log", {"type": "RESULT", "text": f"[{_ts()}]   └─ Density: {pred['density']} g/cm³"})

    # ── Phase 7: Document Generation (Gemini API) ──────────
    yield _event("phase", {"phase": "documenting"})
    await asyncio.sleep(0.3)
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ▶ Phase 7: Generating documents via Gemini AI..."})
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] Calling Gemini 2.0 Flash API..."})

    sop_doc = await generate_sop(params, optimized)
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ SOP generated ({len(sop_doc)} chars)"})

    report_doc = await generate_report(params, optimized)
    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ Report draft generated ({len(report_doc)} chars)"})

    # ── Save to database ───────────────────────────────────
    simulation = Simulation(
        params=params,
        formulation=optimized["formulation"],
        predicted_properties=pred,
        catalyst_name=optimized["selectedCatalyst"]["name"],
        catalyst_type=optimized["selectedCatalyst"]["type"],
        score=optimized["score"],
        msds_warnings_count=len(warnings),
        sop_document=sop_doc,
        report_document=report_doc,
    )
    db.add(simulation)
    await db.commit()
    await db.refresh(simulation)

    yield _event("log", {"type": "EXEC", "text": f"[{_ts()}] ✓ Results saved to database (ID: {simulation.id[:8]}...)"})

    # ── Complete ───────────────────────────────────────────
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ═══════════════════════════════════════"})
    yield _event("log", {"type": "SYSTEM", "text": f"[{_ts()}] ✅ Agent pipeline complete. All 7 phases finished."})
    yield _event("phase", {"phase": "complete"})

    # Send final result
    final_result = {
        "id": simulation.id,
        "formulation": optimized["formulation"],
        "predictedProperties": pred,
        "selectedCatalyst": {
            "name": optimized["selectedCatalyst"]["name"],
            "type": optimized["selectedCatalyst"]["type"],
        },
        "score": optimized["score"],
        "msdsWarnings": warnings,
        "sopDocument": sop_doc,
        "reportDocument": report_doc,
    }
    yield _event("result", final_result)

File: src/engine/agentEngine.js
// ============================================================
// Agent Reasoning Engine - Core Pipeline
// Simulates AI agent's autonomous reasoning and analysis process
// ============================================================

import chemicals from "../data/chemicals.js";
import catalysts from "../data/catalysts.js";
import msdsData from "../data/msds.js";
import { optimizeFormulation } from "./optimizer.js";
import { generateSOP } from "./sopGenerator.js";

/**
 * Delay helper for simulating async processing
 */
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Generate a timestamp string for log entries
 */
const timestamp = () => {
  const now = new Date();
  return now.toLocaleTimeString("en-US", { hour12: false, fractionalSecondDigits: 3 });
};

/**
 * Main Agent Pipeline
 * Runs through 7 sequential reasoning phases, emitting logs via callback
 *
 * @param {Object} params - User input parameters
 * @param {Function} onLog - Callback for streaming log messages
 * @param {Function} onPhaseChange - Callback for phase transitions
 * @returns {Object} Final result with formulation, predictions, SOP
 */
export async function runAgentPipeline(params, onLog, onPhaseChange) {
  const result = {
    selectedMaterials: [],
    selectedCatalyst: null,
    formulation: [],
    predictedProperties: {},
    msdsWarnings: [],
    sopDocument: "",
    reportDocument: "",
    score: 0,
  };

  try {
    // ── Phase 1: Input Parsing ─────────────────────────────
    onPhaseChange("parsing");
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Agent initialized. Starting autonomous design pipeline...` });
    await delay(400);
    onLog({ type: "THINK", text: `[${timestamp()}] Parsing input requirements...` });
    await delay(300);
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Target Heat Resistance: ${params.heatResistance}°C` });
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Target Tensile Strength: ${params.tensileStrength} MPa` });
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Min Eco Score: ${params.ecoScore}` });
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Application Domain: ${params.domain}` });
    onLog({ type: "THINK", text: `[${timestamp()}]   └─ Max Hazard Level: ${params.maxHazard}` });
    await delay(500);
    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ Input validation passed. Constraints loaded.` });

    // ── Phase 2: Database Scan ─────────────────────────────
    onPhaseChange("scanning");
    await delay(600);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 2: Scanning chemical database...` });
    await delay(300);
    onLog({ type: "EXEC", text: `[${timestamp()}] Loading ${chemicals.length} materials from virtual database...` });
    await delay(400);

    // Filter candidates by hazard level
    const hazardRank = { low: 1, medium: 2, high: 3 };
    const maxHazardRank = hazardRank[params.maxHazard] || 2;

    const candidates = chemicals.filter((c) => {
      return hazardRank[c.hazardLevel] <= maxHazardRank && c.category !== "Additive";
    });

    onLog({ type: "THINK", text: `[${timestamp()}] Filtering by hazard constraint (≤ ${params.maxHazard})...` });
    await delay(300);
    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ ${candidates.length} candidate materials passed hazard filter.` });
    await delay(200);

    // Log some candidates
    for (let i = 0; i < Math.min(5, candidates.length); i++) {
      onLog({
        type: "THINK",
        text: `[${timestamp()}]   ${i === Math.min(5, candidates.length) - 1 ? "└─" : "├─"} [${candidates[i].id}] ${candidates[i].name} | Heat: ${candidates[i].heatResistance}°C | Strength: ${candidates[i].tensileStrength} MPa`,
      });
      await delay(150);
    }
    if (candidates.length > 5) {
      onLog({ type: "THINK", text: `[${timestamp()}]      ... and ${candidates.length - 5} more candidates` });
    }

    // ── Phase 3: Compatibility Analysis ────────────────────
    onPhaseChange("analyzing");
    await delay(500);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 3: Catalyst-polymer compatibility analysis...` });
    await delay(400);

    // Find compatible catalysts
    const compatibleCatalysts = catalysts.filter((cat) => {
      return candidates.some((c) => c.compatibleWith.includes(cat.id));
    });

    for (const cat of compatibleCatalysts.slice(0, 4)) {
      const compatCount = candidates.filter((c) => c.compatibleWith.includes(cat.id)).length;
      onLog({
        type: "THINK",
        text: `[${timestamp()}]   ├─ ${cat.name}: compatible with ${compatCount}/${candidates.length} candidates | Selectivity: ${cat.selectivity}%`,
      });
      await delay(200);
    }

    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ ${compatibleCatalysts.length} compatible catalysts identified.` });

    // ── Phase 4: MSDS Safety Review ────────────────────────
    onPhaseChange("safety");
    await delay(500);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 4: MSDS safety review...` });
    await delay(400);

    const warnings = [];
    for (const cand of candidates) {
      const msds = msdsData.find((m) => m.chemicalId === cand.id);
      if (msds) {
        if (msds.signalWord === "Danger") {
          warnings.push({
            materialId: cand.id,
            materialName: cand.name,
            signalWord: msds.signalWord,
            hazards: msds.hazardStatements.slice(0, 2),
            protection: msds.personalProtection,
          });
          onLog({
            type: "WARN",
            text: `[${timestamp()}] ⚠ MSDS ALERT: ${cand.name} — Signal Word: DANGER`,
          });
          await delay(200);
          onLog({
            type: "WARN",
            text: `[${timestamp()}]   └─ Required PPE: ${msds.personalProtection}`,
          });
          await delay(150);
        } else {
          onLog({
            type: "THINK",
            text: `[${timestamp()}]   ├─ ${cand.name}: MSDS reviewed — ${msds.hazardClassification}`,
          });
          await delay(100);
        }
      }
    }

    result.msdsWarnings = warnings;
    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ Safety review complete. ${warnings.length} hazard alert(s) flagged.` });

    // ── Phase 5: Formulation Optimization ──────────────────
    onPhaseChange("optimizing");
    await delay(600);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 5: Running multi-objective optimization...` });
    await delay(300);
    onLog({ type: "EXEC", text: `[${timestamp()}] Executing optimization script...` });
    onLog({ type: "EXEC", text: `[${timestamp()}] > python3 sim_optimize.py --mode=pareto --generations=500` });
    await delay(500);

    const optimized = optimizeFormulation(params, candidates, compatibleCatalysts);

    onLog({ type: "THINK", text: `[${timestamp()}] Weighted Euclidean distance minimization...` });
    await delay(300);
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Generation 100/500: best_fitness = ${(optimized.score * 0.7).toFixed(4)}` });
    await delay(200);
    onLog({ type: "THINK", text: `[${timestamp()}]   ├─ Generation 300/500: best_fitness = ${(optimized.score * 0.9).toFixed(4)}` });
    await delay(200);
    onLog({ type: "THINK", text: `[${timestamp()}]   └─ Generation 500/500: best_fitness = ${optimized.score.toFixed(4)} ★ converged` });
    await delay(300);

    result.selectedMaterials = optimized.selectedMaterials;
    result.selectedCatalyst = optimized.selectedCatalyst;
    result.formulation = optimized.formulation;
    result.score = optimized.score;

    onLog({ type: "RESULT", text: `[${timestamp()}] ═══════════════════════════════════════` });
    onLog({ type: "RESULT", text: `[${timestamp()}] ★ Optimal formulation found (score: ${optimized.score.toFixed(2)}/100)` });
    for (const f of optimized.formulation) {
      onLog({
        type: "RESULT",
        text: `[${timestamp()}]   ├─ ${f.name}: ${f.ratio.toFixed(1)}%`,
      });
      await delay(100);
    }
    onLog({
      type: "RESULT",
      text: `[${timestamp()}]   └─ Catalyst: ${optimized.selectedCatalyst.name}`,
    });

    // ── Phase 6: Property Prediction ───────────────────────
    onPhaseChange("simulating");
    await delay(500);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 6: Predicting composite properties...` });
    await delay(300);
    onLog({ type: "EXEC", text: `[${timestamp()}] > python3 property_predictor.py --model=neural_net_v3` });
    await delay(600);

    // Calculate predicted properties via weighted average
    const predicted = {
      heatResistance: 0,
      tensileStrength: 0,
      elongation: 0,
      ecoScore: 0,
      density: 0,
    };

    for (const f of optimized.formulation) {
      const mat = chemicals.find((c) => c.id === f.id);
      if (mat) {
        const w = f.ratio / 100;
        predicted.heatResistance += mat.heatResistance * w;
        predicted.tensileStrength += mat.tensileStrength * w;
        predicted.elongation += mat.elongation * w;
        predicted.ecoScore += mat.ecoScore * w;
        predicted.density += mat.density * w;
      }
    }

    // Apply catalyst effects
    const cat = optimized.selectedCatalyst;
    predicted.heatResistance *= cat.effectOnHeatRes;
    predicted.tensileStrength *= cat.effectOnStrength;
    predicted.ecoScore *= cat.effectOnEco;

    // Round values
    Object.keys(predicted).forEach((key) => {
      predicted[key] = Math.round(predicted[key] * 10) / 10;
    });

    result.predictedProperties = predicted;

    onLog({ type: "RESULT", text: `[${timestamp()}] ── Predicted Composite Properties ──` });
    onLog({ type: "RESULT", text: `[${timestamp()}]   ├─ Heat Resistance: ${predicted.heatResistance}°C (target: ${params.heatResistance}°C)` });
    onLog({ type: "RESULT", text: `[${timestamp()}]   ├─ Tensile Strength: ${predicted.tensileStrength} MPa (target: ${params.tensileStrength} MPa)` });
    onLog({ type: "RESULT", text: `[${timestamp()}]   ├─ Elongation: ${predicted.elongation}%` });
    onLog({ type: "RESULT", text: `[${timestamp()}]   ├─ Eco Score: ${predicted.ecoScore}/100` });
    onLog({ type: "RESULT", text: `[${timestamp()}]   └─ Density: ${predicted.density} g/cm³` });

    // ── Phase 7: SOP Generation ────────────────────────────
    onPhaseChange("documenting");
    await delay(500);
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ▶ Phase 7: Generating SOP & Report documents...` });
    await delay(400);
    onLog({ type: "EXEC", text: `[${timestamp()}] Compiling Standard Operating Procedure...` });
    await delay(600);

    const sopDoc = generateSOP(params, result);
    result.sopDocument = sopDoc.sop;
    result.reportDocument = sopDoc.report;

    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ SOP document generated (${sopDoc.sop.length} chars)` });
    await delay(200);
    onLog({ type: "EXEC", text: `[${timestamp()}] ✓ Report draft generated (${sopDoc.report.length} chars)` });
    await delay(300);

    onLog({ type: "SYSTEM", text: `[${timestamp()}] ═══════════════════════════════════════` });
    onLog({ type: "SYSTEM", text: `[${timestamp()}] ✅ Agent pipeline complete. All phases finished successfully.` });
    onLog({ type: "SYSTEM", text: `[${timestamp()}] Total materials analyzed: ${chemicals.length}` });
    onLog({ type: "SYSTEM", text: `[${timestamp()}] Optimization score: ${result.score.toFixed(2)}/100` });

    onPhaseChange("complete");
    return result;
  } catch (err) {
    onLog({ type: "ERROR", text: `[${timestamp()}] ✗ Agent error: ${err.message}` });
    onPhaseChange("error");
    throw err;
  }
}
